pkg/asset/storage.go

```diff
@@ -0,0 +1,369 @@
+}
+
+func (storage *Storage) UploadBuildAssetCopy(origFile string, assetType Type, copyName string,
+	build *dashapi.Build) error {
+	dir, err := ioutil.TempDir("", "")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+	tmpFile := filepath.Join(dir, copyName)
+	if err := osutil.CopyFile(origFile, tmpFile); err != nil {
+		return fmt.Errorf("failed to copy the file: %w", err)
+	}
+	defer os.Remove(tmpFile)
+	return storage.UploadBuildAsset(tmpFile, assetType, build)
+}
+
+var ErrAssetDoesNotExist = errors.New("the asset did not exist")
+
+const deletionEmbargo = time.Hour * 24 * 7
+
+func (storage *Storage) DeprecateAssets() error {
+	resp, err := storage.dash.NeededAssetsList()
+	if err != nil {
+		return fmt.Errorf("failed to query needed assets: %w", err)
+	}
+	needed := map[string]bool{}
+	for _, url := range resp.DownloadURLs {
+		needed[url] = true
+	}
+	storage.tracer.Log("queried needed assets URLs: %#v", needed)
+	existing, err := storage.backend.list()
+	if err != nil {
+		return fmt.Errorf("failed to query object list: %w", err)
+	}
+	toDelete := []string{}
+	intersection := 0
+	for _, obj := range existing {
+		keep := false
+		if time.Since(obj.createdAt) < deletionEmbargo {
+			// To avoid races between object upload and object deletion, we don't delete
+			// newly uploaded files for a while after they're uploaded.
+			keep = true
+		}
+		if val, ok := needed[obj.downloadURL]; ok && val {
+			keep = true
+			intersection++
+		}
+		storage.tracer.Log("-- object %v, %v: keep %t", obj.downloadURL, obj.createdAt, keep)
+		if !keep {
+			toDelete = append(toDelete, obj.downloadURL)
+		}
+	}
+	const intersectionCheckCutOff = 4
+	if len(existing) > intersectionCheckCutOff && intersection == 0 {
+		// This is a last-resort protection against possible dashboard bugs.
+		// If the needed assets have no intersection with the existing assets,
+		// don't delete anything. Otherwise, if it was a bug, we will lose all files.
+		return fmt.Errorf("needed assets have some intersection with the existing ones")
+	}
+	for _, url := range toDelete {
+		err := storage.backend.remove(url)
+		storage.tracer.Log("-- deleted %v: %v", url, err)
+		// Several syz-ci's might be sharing the same storage. So let's tolerate
+		// races during file deletion.
+		if err != nil && err != ErrAssetDoesNotExist {
+			return fmt.Errorf("asset deletion failure: %w", err)
+		}
+	}
+	return nil
+}
+
+type uploadRequest struct {
+	origFile        string
+	savePath        string
+	contentEncoding string
+	contentType     string
+}
+
+type uploadResponse struct {
+	downloadURL string
+}
+
+type storedObject struct {
+	downloadURL string
+	createdAt   time.Time
+}
+
+type StorageBackend interface {
+	upload(req *uploadRequest) (*uploadResponse, error)
+	list() ([]storedObject, error)
+	remove(path string) error
+}
+
+type PreprocessorCollection struct {
+	defaultOne Preprocessor
+	custom     map[Type]Preprocessor
+	// Map cannot be concurrently accessed.
```

Comment:
We don't change it after creation so we don't need any protection. Reading from the map can be done concurrnetly.